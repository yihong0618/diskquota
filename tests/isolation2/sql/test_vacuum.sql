-- This file tests various race conditions when performing 'VACUUM FULL'.

-- 1. When the gpdb is performing 'VACUUM FULL' on some relation, it can be summarized
--    as the following 3 steps:
--    s1) create a new temporary relation (smgrcreate hook will be triggered, newly
--        created relfilenode will be put into shmem).
--    s2) insert data into the newly created relation from the old relation (smgrextend
--        hook will be triggered, newly created relfilenode will be put into shmem).
--    s3) change the old relation's relfilenode to the newly created one.
--    Consider the following situation:
--    If the diskquota bgworker pulls active oids before the 'VACUUM FULL' operation finishing,
--    the newly created relfilenode is translated to the newly created temporary relation's oid,
--    the old relation's size cannot be updated. We resolve it by making altered relations' oids
--    constantly active so that the diskquota bgworker keeps updating the altered relation size
--    during 'VACUUM FULL'.
CREATE TABLE dummy_t1(i int) DISTRIBUTED BY (i);
INSERT INTO dummy_t1 SELECT generate_series(1, 1000);
DELETE FROM dummy_t1;
-- Wait for the diskquota bgworker refreshing the size of 'dummy_t1'.
SELECT diskquota.wait_for_worker_new_epoch();
-- Shows that the result of pg_table_size() and diskquota.table_size are identical.
SELECT pg_table_size('dummy_t1');
SELECT tableid::regclass, size, segid FROM diskquota.table_size
  WHERE tableid='dummy_t1'::regclass ORDER BY segid;
SELECT gp_inject_fault_infinite('object_access_post_alter', 'suspend', dbid)
  FROM gp_segment_configuration WHERE role='p' AND content<>-1;
SELECT diskquota.wait_for_worker_new_epoch();
1&: VACUUM FULL dummy_t1;
SELECT gp_wait_until_triggered_fault('object_access_post_alter', 1, dbid)
  FROM gp_segment_configuration WHERE role='p' AND content<>-1;
-- Wait for the diskquota bgworker 'consumes' the newly created relfilenode from shmem.
SELECT diskquota.wait_for_worker_new_epoch();
SELECT gp_inject_fault_infinite('object_access_post_alter', 'reset', dbid)
  FROM gp_segment_configuration WHERE role='p' AND content<>-1;
1<:

-- Wait for the diskquota bgworker refreshing the size of 'dummy_t1'.
SELECT diskquota.wait_for_worker_new_epoch();
-- Shows that the result of pg_table_size() and diskquota.table_size are identical.
SELECT pg_table_size('dummy_t1');
SELECT tableid::regclass, size, segid FROM diskquota.table_size
  WHERE tableid='dummy_t1'::regclass ORDER BY segid;
DROP TABLE dummy_t1;
