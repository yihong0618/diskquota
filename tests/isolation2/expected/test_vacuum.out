-- This file tests various race conditions when performing 'VACUUM FULL'.

-- 1. When the gpdb is performing 'VACUUM FULL' on some relation, it can be summarized
--    as the following 3 steps:
--    s1) create a new temporary relation (smgrcreate hook will be triggered, newly
--        created relfilenode will be put into shmem).
--    s2) insert data into the newly created relation from the old relation (smgrextend
--        hook will be triggered, newly created relfilenode will be put into shmem).
--    s3) change the old relation's relfilenode to the newly created one.
--    Consider the following situation:
--    If the diskquota bgworker pulls active oids before the 'VACUUM FULL' operation finishing,
--    the newly created relfilenode is translated to the newly created temporary relation's oid,
--    the old relation's size cannot be updated. We resolve it by making altered relations' oids
--    constantly active so that the diskquota bgworker keeps updating the altered relation size
--    during 'VACUUM FULL'.
CREATE TABLE dummy_t1(i int) DISTRIBUTED BY (i);
CREATE
INSERT INTO dummy_t1 SELECT generate_series(1, 1000);
INSERT 1000
DELETE FROM dummy_t1;
DELETE 1000
-- Wait for the diskquota bgworker refreshing the size of 'dummy_t1'.
SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t                         
(1 row)
-- Shows that the result of pg_table_size() and diskquota.table_size are identical.
SELECT pg_table_size('dummy_t1');
 pg_table_size 
---------------
 98304         
(1 row)
SELECT tableid::regclass, size, segid FROM diskquota.table_size WHERE tableid='dummy_t1'::regclass ORDER BY segid;
 tableid  | size  | segid 
----------+-------+-------
 dummy_t1 | 98304 | -1    
 dummy_t1 | 32768 | 0     
 dummy_t1 | 32768 | 1     
 dummy_t1 | 32768 | 2     
(4 rows)
SELECT gp_inject_fault_infinite('object_access_post_alter', 'suspend', dbid) FROM gp_segment_configuration WHERE role='p' AND content<>-1;
 gp_inject_fault_infinite 
--------------------------
 Success:                 
 Success:                 
 Success:                 
(3 rows)
SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t                         
(1 row)
1&: VACUUM FULL dummy_t1;  <waiting ...>
SELECT gp_wait_until_triggered_fault('object_access_post_alter', 1, dbid) FROM gp_segment_configuration WHERE role='p' AND content<>-1;
 gp_wait_until_triggered_fault 
-------------------------------
 Success:                      
 Success:                      
 Success:                      
(3 rows)
-- Wait for the diskquota bgworker 'consumes' the newly created relfilenode from shmem.
SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t                         
(1 row)
SELECT gp_inject_fault_infinite('object_access_post_alter', 'reset', dbid) FROM gp_segment_configuration WHERE role='p' AND content<>-1;
 gp_inject_fault_infinite 
--------------------------
 Success:                 
 Success:                 
 Success:                 
(3 rows)
1<:  <... completed>
VACUUM

-- Wait for the diskquota bgworker refreshing the size of 'dummy_t1'.
SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t                         
(1 row)
-- Shows that the result of pg_table_size() and diskquota.table_size are identical.
SELECT pg_table_size('dummy_t1');
 pg_table_size 
---------------
 0             
(1 row)
SELECT tableid::regclass, size, segid FROM diskquota.table_size WHERE tableid='dummy_t1'::regclass ORDER BY segid;
 tableid  | size | segid 
----------+------+-------
 dummy_t1 | 0    | -1    
 dummy_t1 | 0    | 0     
 dummy_t1 | 0    | 1     
 dummy_t1 | 0    | 2     
(4 rows)
DROP TABLE dummy_t1;
DROP
